### PR TITLE
Upgrade Cassandra to Java 11

### PR DESCRIPTION
**Pull request by Sam-Kramer**
Upgrade Cassandra to Java 11

### CODE REVIEW

File: src/java/org/apache/cassandra/security/SSLFactory.java [hunk 1]
```diff
@@ -215,9 +205,18 @@ static SSLSocket maybeConnectWithTimeout(Callable<SSLSocket> callable, long time
...
-    private static void maybeAddSni(InetAddress addr, SSLParameters sslParameters)
+    private static void maybeAddSni(InetAddress addr, SSLParameters sslParameters, EncryptionOptions options)
     {
-        String providedName = addr.getHostName();
+        Set<String> validServerNames = getValidServerNames(options);
+        String hostname = addr.getHostName();
+
+        if(!validServerNames.isEmpty() && !validServerNames.contains(hostname)) {
+            hostname = addr.getCanonicalHostName();
+            logger.info("Using canoical hostname {} instead of localhost {} as no entry for hostname found in keystore {}", addr.getHostName(), hostname, validServerNames);
```
Comment (Sam-Kramer) [inline src/java/org/apache/cassandra/security/SSLFactory.java]: typo

File: src/java/org/apache/cassandra/utils/FBUtilities.java [hunk 2]
```diff
@@ -133,7 +133,7 @@ public static InetAddress getLocalAddress()
             try
             {
                 localInetAddress = DatabaseDescriptor.getListenAddress() == null
-                                    ? InetAddress.getLocalHost()
+                                    ? InetAddress.getByName(InetAddress.getLocalHost().getCanonicalHostName())
```
Comment (leonz) [inline src/java/org/apache/cassandra/utils/FBUtilities.java:136]: This generally seems reasonable.  How does this resolve in Kubernetes?

Comment (Sam-Kramer) [reply src/java/org/apache/cassandra/utils/FBUtilities.java:136]: That's a good flag. I tested this myself, and we should be ok here.

```
ptdocker@cassandra-0:/mnt/secrets/certs$ hostname -f 
cassandra-0.cassandra.<ns>.svc.cluster.local
```
Interesting, `hostname` returns just `cassandra-0`, but the certs are much more resilient in k8s, where they do all permutations of the FQDN.

Comment (Sam-Kramer) [reply src/java/org/apache/cassandra/utils/FBUtilities.java:136]: Actually that's a bit of a lie -- in fact, this fix is also required for K8s, as there is no entry for just the hostname!

**Review by leonz — APPROVED** [review]

**Pull request merged**